### PR TITLE
Link against RMI native code when RMI is used

### PR DESCRIPTION
**Pull request by koutheir**
Link against RMI native code when RMI is used.

### CODE REVIEW

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationRMI.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+    private static final String STATIC_LIB = "rmi";
+    private static final String GC = "sun.rmi.transport.GC";
+    private static final String GC_LOADER = "sun.rmi.transport.GC$1";
+
+    @Override
+    public void beforeAnalysis(BeforeAnalysisAccess access) {
+        /*
+         * The default RMI class implementations trigger loading of the 'rmi' library in their
+         * static initializers on Java 11. We force a class initialization rerun in order to ensure
```
Comment (fniephaus) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationRMI.java:51]: Not just on Java 11, but on Java 11+: https://github.com/openjdk/jdk/blob/52388179e65d4703ec33569dcc7c1351c57e6056/src/java.rmi/share/classes/sun/rmi/transport/GC.java#L90

Comment (koutheir) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationRMI.java:51]: I didn't check other Java versions, as we only use Java 11. What is the change needed here?

Comment (fniephaus) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationRMI.java:51]: I checked whether rmi is still loaded in the static initializer in the JDK main sources and thought I leave a comment. We could extend the Java comment when we integrate this:
```suggestion
         * static initializers on Java 11+. We force a class initialization rerun in order to ensure
```

Comment (koutheir) [reply substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationRMI.java]: Done.

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationRMI.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+         * static initializers on Java 11+. We force a class initialization rerun in order to ensure
+         * we pick up the loadLibrary call and properly link against the library.
+         */
+        rerunClassInit(access, GC, GC_LOADER);
+
+        ArrayList<Class<?>> triggers = new ArrayList<>();
+        triggers.add(clazz(access, GC));
+        triggers.add(clazz(access, GC_LOADER));
+
+        access.registerReachabilityHandler(JNIRegistrationRMI::handleGCClassReachable, triggers.toArray());
```
Comment (fniephaus) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationRMI.java:60]:
```suggestion
        access.registerReachabilityHandler(JNIRegistrationRMI::handleGCClassReachable, clazz(access, GC), clazz(access, GC_LOADER));
```

File: substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationRMI.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    private static void handleGCClassReachable(@SuppressWarnings("unused") DuringAnalysisAccess access) {
+        NativeLibraries nativeLibraries = ((FeatureImpl.DuringAnalysisAccessImpl) access).getNativeLibraries();
+
+        NativeLibrarySupport.singleton().preregisterUninitializedBuiltinLibrary(STATIC_LIB);
+        nativeLibraries.addStaticJniLibrary(STATIC_LIB);
```
Comment (fniephaus) [inline substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/jdk/JNIRegistrationRMI.java:67]:
```suggestion
    private static void handleGCClassReachable(DuringAnalysisAccess access) {
        NativeLibrarySupport.singleton().preregisterUninitializedBuiltinLibrary(STATIC_LIB);
        NativeLibraries nativeLibraries = ((FeatureImpl.DuringAnalysisAccessImpl) access).getNativeLibraries();
        nativeLibraries.addStaticJniLibrary(STATIC_LIB);
```

**Review by olpaw — APPROVED** [review]
Now with the suggestions from @fniephaus this LGTM

**Comment by olpaw** [comment]
> I don't think we need this PR anymore, because we don't need the RMI library: https://github.com/oracle/graal/pull/6250.

Right. @koutheir with @fniephaus changes your PR seems obsolete.

**Pull request closed without merging**